lib/facebook_ads/model.rb

```
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+require 'facebook_ads/json_converter'
+
+module FacebookAds
+  class Model
+    class << self
+      def field_converters
+        @field_converters ||= {}
+      end
+
+      def known_fields
+        field_converters.keys
+      end
+
+      def field(name, type: nil)
+        name = name.to_s
+        field_converters[name] = JsonConverter.for(type)
+
+        define_method name do
+          fields[name]
+        end
+
+        define_method "#{name}=" do |value|
+          fields[name] = value
+        end
+      end
+    end
+
+    attr_reader :id, :fields, :client
+
+    def initialize(id, client: FacebookAds.client)
+      @id = id
+      @fields = {}
+      @client = client
+    end
+
+    def fetch(fields)
+      fields = Array(fields)
+      validate_fields(fields)
+
+      response = client.get(id, params: { fields: fields.join(',') })
+      response.each do |field, value|
+        converter = self.class.field_converters[field]
+        next if converter.nil?
```

How can this happen? Only way I see is if someone removes a Converter on purpose. And I think we should raise an error if a Converter is missing instead of just skipping the field.

---

It can happen when the response contains a field that we did not expect.

Since we already make sure that we can only request fields that we know, this can only happen if FB decides to answer with more fields than we asked for.

I know that they do this for the `id` field, so this is one that we are skipping in practice.
Leaving the `id` aside (because I could fix that in a different way), I am not sure if there are other cases where they respond with fields we did not request.

I would not consider those cases errors, however. Since it is only additional information.

Thinking about it: We should probably compare the fields we received with the fields we requested. Because it would be an error, if one was **missing** from the response.

---

Okay that explanation makes sense and I'm fine with simply skipping fields that were not requested.
I do agree that a missing field is an error. I'm unsure whether we should check for it just because it could be missing. If it happens and no error is raised it will be very annoying. On the other hand. I don't think that an API simply not providing a requested field (without raising an error) is rather unlikely and checking for it might a be a case of YAGNI. So... what do you think? Better safe than sorry?

---

I'd prefer to keep it simple now and start to hate the API in the future, when we find out that it behaves in this annoying way... (and then do the check of course)

---

Agreed.